### ISSUE TITLE
Establish the connection between sent and acknowledged messages in producer

### ISSUE THREAD

**Issue by enchantner** [confluentinc/confluent-kafka-go#330]
Description
===========

This is more of a question rather than an issue. I know I'm kinda not supposed to want this, but...

The messages I'm producing have different logical groups, and I have a number of messages to be sent which have the same Group ID, though input data is not sorted by it.

I need to determine exactly when I've sent ALL messages I needed (I have exact numbers for each group) with the same Group ID, to be able to execute some logic after that, making sure it's consistent.

Is it even possible to do such a thing in an asynchronous fashion?

How to reproduce
================

My producer, following the example, consists basically of two methods. The one for sending the data:

`func (kfp *KafkaProducer) ProduceData(dataCh <-chan *types.Chunk) {
	defer kfp.wg.Done()

	for chunk := range dataCh {
		kfp.producer.ProduceChannel() <- &kafka.Message{
			TopicPartition: kafka.TopicPartition{
				Topic:     &kfp.Topic,
				Partition: kafka.PartitionAny},
			Value: chunk.Payload}

		select {
		case <-kfp.stopCh:
			return
		default:
			// continue working
		}
	}
}`

...and the one to process the incoming "acks":

`func (kfp *KafkaProducer) Poll() {
	run := true
	for run {
		select {
		case <-kfp.stopCh:
			run = false
		case ev := <-kfp.producer.Events():
			switch m := ev.(type) {
			case *kafka.Message:
				if m.TopicPartition.Error != nil {
					logger.Info.Printf("Delivery failed: %v\n", m.TopicPartition.Error)
				} else {
                                       // here I'll need some logic to check if all chunks in a group 
                                       // are successfully sent, and execute some logic if that's the case
				}
			default:
				logger.Info.Printf("Ignored event: %s\n", ev)
			}
		}
	}
}`

I've removed some utility parts of the code on counting messages and determining the loop brake, because it doesn't really matter here. What matters is, can I somehow determine which messages got acknowledged, in order to keep track of the chunks' group? Maybe I can store Group ID in some field inside message object (which doesn't affect the serialized payload) and check it afterwards?

**Comment by edenhill** [comment]
This is what the `Message.Opaque` field is for; allowing your application associate an arbitrary application object with the Message object, which will be available in the delivery report event.

**Comment by edenhill** [comment]
Do note that Message.Opaque is a client-local field, it is not sent to the Kafka broker.

**Comment by enchantner** [comment]
Could you please provide an example on using that? Currently I'm assigning a pointer to it, and on the delivery I'm casting Opaque to that pointer type, but getting `panic: interface conversion: interface {} is nil, not *types.Feed`.

**Comment by edenhill** [comment]
@enchantner Please provide relevant parts of your code

**Comment by enchantner** [comment]
So, I'm sending it like this:

```
func (kfp *KafkaProducer) ProduceData(dataCh <-chan *types.Chunk) {
	defer kfp.wg.Done()

	for chunk := range dataCh {
		kfp.producer.ProduceChannel() <- &kafka.Message{
			TopicPartition: kafka.TopicPartition{
				Topic:     &kfp.Topic,
				Partition: kafka.PartitionAny},
			Value: chunk.Payload,
                        Opaque: chunk.Diff.Feed}

		select {
		case <-kfp.stopCh:
			return
		default:
			// continue working
		}
	}
}
```

And receiving like this:
```
func (kfp *KafkaProducer) Poll() {
	run := true
	for run {
		select {
		case <-kfp.stopCh:
			run = false
		case ev := <-kfp.producer.Events():
			switch m := ev.(type) {
			case *kafka.Message:
				if m.TopicPartition.Error != nil {
					logger.Info.Printf("Delivery failed: %v\n", m.TopicPartition.Error)
				} else {
                                       initialFeed := m.Opaque.(*types.Feed)
                                       // do stuff with *types.Feed object
				}
			default:
				logger.Info.Printf("Ignored event: %s\n", ev)
			}
		}
	}
}
```

**Comment by edenhill** [comment]
Are you setting `go.batch.producer`? If so, that might not work with message Opaque

**Comment by enchantner** [comment]
Yes, looks like that's the issue. Removed that setting and it works. Thanks!

I'm just curious though, why is it a problem to store that in case of batch producer? Also, are there any downsides of disabling it?

**Comment by edenhill** [comment]
It is just a lapse in the implementation, I believe passing the opaque should be technically possible.
The batch producer is currently a bit problematic since it can't be used to represent all fields in a message, such as headers and timestamp.
Most users will not need to use it though, only when you're in need of maximum thruput.
